src/spyglass/spikesorting/v1/recording.py

```diff
@@ -0,0 +1,730 @@
+                            0
+                        ]
+                    else:
+                        ValueError(
+                            f"Error in electrode group {e_group}: reference electrodes are not all the same"
+                        )
+                else:
+                    if e_group not in references.keys():
+                        raise Exception(
+                            f"electrode group {e_group} not a key in references, so cannot set reference"
+                        )
+                    else:
+                        sg_key["sort_reference_electrode_id"] = references[
+                            e_group
+                        ]
+                # Insert sort group and sort group electrodes
+                reference_electrode_group = electrodes[
+                    electrodes["electrode_id"]
+                    == sg_key["sort_reference_electrode_id"]
+                ][
+                    "electrode_group_name"
+                ]  # reference for this electrode group
+                if (
+                    len(reference_electrode_group) == 1
+                ):  # unpack single reference
+                    reference_electrode_group = reference_electrode_group[0]
+                elif (int(sg_key["sort_reference_electrode_id"]) > 0) and (
+                    len(reference_electrode_group) != 1
+                ):
+                    raise Exception(
+                        f"Should have found exactly one electrode group for reference electrode,"
+                        f"but found {len(reference_electrode_group)}."
+                    )
+                if omit_ref_electrode_group and (
+                    str(e_group) == str(reference_electrode_group)
+                ):
+                    print(
+                        f"Omitting electrode group {e_group} from sort groups because contains reference."
+                    )
+                    continue
+                shank_elect = electrodes["electrode_id"][
+                    np.logical_and(
+                        electrodes["electrode_group_name"] == e_group,
+                        electrodes["probe_shank"] == shank,
+                    )
+                ]
+                if (
+                    omit_unitrode and len(shank_elect) == 1
+                ):  # omit unitrodes if indicated
+                    print(
+                        f"Omitting electrode group {e_group}, shank {shank} from sort groups because unitrode."
+                    )
+                    continue
+                cls.insert1(sg_key, skip_duplicates=True)
+                for elect in shank_elect:
+                    sge_key["electrode_id"] = elect
+                    cls.SortGroupElectrode.insert1(
+                        sge_key, skip_duplicates=True
+                    )
+                sort_group += 1
+
+
+@schema
+class SpikeSortingPreprocessingParameter(dj.Lookup):
```

Comment:
Just a style question, but why have singular for parameter? It looks like most the other pipelines use plural

Reply:
Are you asking why Parameter tables specifically are singular or why table names in general (in this pipeline) are singular?

I made this decision thinking that the convention recommended by datajoint is to make table names singular. I thought that was in the [documentation](https://datajoint.com/docs/core/datajoint-python/0.14/design/tables/declare/)... but can no longer find it, so I may have hallucinated that! (@CBroz1 was there ever such a convention?) I don't have a strong opinion about this but figured it would be easier if every table name was singular since most table we already have are singular.

Reply:
I vaguely recall that too - but it might have been a stylistic preference Dimitri expressed verbally. I'd prioritize alignment with existing tables. Doing a global grep on the repo (`class .*Param.*\(dj`), I'm seeing plural `Parameters`, or more recent `Params`

Reply:
Switching back to `Parameters`

Reply:
Should the keys also be plural like they were before?

Reply:
yes switched the key `preproc_param` back to `preproc_params`.